packages/ts/file-router/src/runtime/RouterBuilder.ts

```diff
@@ -0,0 +1,122 @@
+import { protectRoutes, type RouteObjectWithAuth } from '@vaadin/hilla-react-auth';
+import { type ComponentType, createElement, type ReactElement } from 'react';
+import { createBrowserRouter, type RouteObject } from 'react-router-dom';
+import { transformTreeSync } from '../shared/transformTree.js';
+import type { AgnosticRoute } from '../types.js';
+import { toReactRouter } from './toReactRouter.js';
+
+/**
+ * Deeply merges two lists of routes. If the specific path is already present,
+ * the route is merged, otherwise the new routes are added to the list.
+ *
+ * @param a - The first list of routes.
+ * @param b - The second list of routes.
+ */
+function mergeRoutes(a: RouteObject[], b: RouteObject[]): RouteObject[] {
+  return b.reduce(
+    (result, route) => {
+      const existingRoute = result.find((r) => r.path === route.path);
+      if (existingRoute) {
+        Object.assign(existingRoute, route);
+        existingRoute.children = existingRoute.children
+          ? mergeRoutes(existingRoute.children, route.children ?? [])
+          : route.children;
+      } else {
+        result.push(route);
+      }
+      return result;
+    },
+    [...a],
+  );
+}
+
+/**
+ * A builder for creating a Vaadin-specific router for React with
+ * authentication and server routes support.
+ */
+export class RouterBuilder {
+  #routes: RouteObject[] = [];
+
+  /**
+   * Returns the current list of routes.
+   */
+  get routes(): readonly RouteObject[] {
+    return this.#routes;
+  }
+
+  /**
+   * Adds the given routes to the current list of routes. All the routes are
+   * deeply merged to preserve the path uniqueness.
+   *
+   * @param routes - A list of routes to add to the current list.
+   */
+  withReactRoutes(...routes: RouteObject[]): this {
+    if (!this.#routes.length) {
+      this.#routes = routes;
+    } else {
+      this.#routes = mergeRoutes(this.#routes, routes);
+    }
+
+    return this;
+  }
+
+  /**
+   * Adds the given file routes to the current list of routes. All the routes
+   * are transformed to React RouterObjects and deeply merged to preserve the
+   * path uniqueness.
+   *
+   * @param routes - A list of routes to add to the current list.
+   */
+  withFileRoutes(...routes: readonly AgnosticRoute[]): this {
+    const reactRoutes = routes.map(toReactRouter);
+    this.withReactRoutes(...reactRoutes);
+    return this;
+  }
+
+  /**
+   * Adds the given server route element to each branch of the current list of
+   * routes.
+   *
+   * @param component - The React component to add to each branch of the
+   * current list of routes.
+   */
+  withServerRoutes(component: ComponentType): this {
```

Comment:
I propose to rename it to `withFallbackComponent`, as it takes a component and might be something else that Flow route component, e.g. custom 404 page/component.

Reply:
Renamed to `withFallback`